web/api/v1/api.go

```diff
@@ -541,7 +541,15 @@ func (api *API) series(r *http.Request) apiFuncResult {
 		matcherSets = append(matcherSets, matchers)
 	}
 
-	q, err := api.Queryable.Querier(r.Context(), timestamp.FromTime(start), timestamp.FromTime(end))
+	// HACK: Timeout querier operations after 60s from now to avoid OOM
+	// for long running queries.
+	//
+	// Fixes issue: https://github.com/prometheus/prometheus/issues/5547
+	// Related issue: https://github.com/prometheus/prometheus/issues/5878
+	ctx, cancel := context.WithTimeout(r.Context(), 60*time.Second)
+	defer cancel()
+
+	q, err := api.Queryable.Querier(ctx, timestamp.FromTime(start), timestamp.FromTime(end))
```

Comment:
Don't we want some `select .. case <-s.ctx.Done()` inside the `for set.Next()` loop on line 571? That's where the memory is growing.

Reply:
Oh never mind I see the check in baseChunkSeries.Next(). Cool lg2m!